### ISSUE TITLE
Extension causes high cpu load

### ISSUE THREAD

**Issue by pradipsilwal** [hashicorp/vscode-terraform#501]
- Issue Type: `Performance`
- Extension Name: `terraform`
- Extension Version: `2.2.3`
- OS Version: `Windows_NT x64 10.0.19041`
- VSCode version: `1.50.1`

:warning: Make sure to **attach** this file from your *home*-directory:
:warning:`c:\Users\silwa\AppData\Local\Temp\hashicorp.terraform-unresponsive.cpuprofile.txt`

Find more details here: https://github.com/microsoft/vscode/wiki/Explain-extension-causes-high-cpu-load

**Comment by kiklop74** [comment]
Same happens to me and reason for this is that upon starting the VS Code extension executes multiple instances of terraform cli trying to obtain information about providers etc. Not sure how to disable this.

**Comment by sidewinder12s** [comment]
Also seeing similar behavior, my environment is setup to open a monorepo with probably 100+ tf backends/workspaces. Wondering if it can more nicely start up/iterate through all those. 

Or if this setting might help?
https://code.visualstudio.com/api/references/activation-events#onStartupFinished

**Comment by snolan-amount** [comment]
I use tfenv to allow for quick switching between terraform versions on a Mac, so different OS but the issue is the same. On opening a folder where there are lots of terraform projects CPU gets very high, then my computer becomes completely unresponsive and then is restarts. It is definitely this extension causing the problem, I can watch the dozens of tfenv and terraform processes running and eventually crashing the computer.
I grabbed a screenshot before the computer crashed but you get the idea. The extension needs to limit the number of concurrent processes that it spawns to run terraform in all the child folders.
![Screen Shot 2020-12-17 at 8 13 43 PM](https://user-images.githubusercontent.com/55757441/102566362-03a72800-40a5-11eb-8231-f964a6c066a6.png)

**Comment by dominik-lekse** [comment]
With the Terraform extension enabled in a workspace with 50+ Terraform root modules, I also experience the OS becoming unresponsive due to a high CPU load when opening the workspace.

I assume that this is related to the method terraform-ls inspects the root modules. There is an open issue for terraform-ls to cap the auto-loading: https://github.com/hashicorp/terraform-ls/issues/186.

My hope is, that terraform-ls allows to configure a limit for the auto-loading in the future. I still do not understand why terraform-ls tries to auto-load the modules in a so massively parallel manner.

**Comment by dominik-lekse** [comment]
For affected workspaces, I am using the following settings as a bad workaround to prevent starting terraform-ls and thus prevent massive slow down of the system while at least keep the syntax highlighting feature of the extension.

```json
{
	"settings": {
		"terraform.languageServer": {
			"external": true,
			// Configure a non-existing path for terraform-ls to prevent massive slow down of the system for workspaces with lots of Terraform modules
			"pathToBinary": "/Users/lekse/bin/terraform-ls-fake",
			"args": [
				"serve"
			],
			"maxNumberOfProblems": 1,
			"trace.server": "off"
		}
	}
}
```

**Comment by radeksimko** [comment]
Hi folks,
What I believe was the main cause of the CPU spikes should be resolved in 0.7.0 and future LS releases as per https://github.com/hashicorp/terraform-ls/pull/391

With that I'm going to close this issue, but feel free to open a new one if you continue having issues and include the version of the language server and the extension in such new report please.
